File: world.go
package main

type world struct {
	name     string
	systems  []runner
	entities []*entity
}

func (w *world) run() {
	for _, system := range w.systems {
		system.run()
	}
}

func (w *world) addEntity(e *entity) {
	w.entities = append(w.entities, e)
}

func (w *world) addSystem(s runner) {
	w.systems = append(w.systems, s)
}

func (w *world) sortEntities() {
	for _, system := range w.systems {
		system.sortEntities()
	}
}

// Update all systems with the entities they will be run on
func (w *world) updateSystems() {
	hasComponents := true
	for _, system := range w.systems {
		for _, entity := range w.entities {
			for _, component := range system.getComponents() {
				switch component {
				case "appearance":
					if entity.appearance == nil {
						hasComponents = false
					}
				case "ui":
					if entity.ui == nil {
						hasComponents = false
					}
				case "camera":
					if entity.camera == nil {
						hasComponents = false
					}
				case "collision":
					if entity.collision == nil {
						hasComponents = false
					}
				case "controls":
					if entity.controls == nil {
						hasComponents = false
					}
				case "velocity":
					if entity.velocity == nil {
						hasComponents = false
					}
				case "gravity":
					if entity.gravity == nil {
						hasComponents = false
					}
				case "position":
					if entity.position == nil {
						hasComponents = false
					}
				}
			}
			if hasComponents {
				system.addEntity(entity)
			}
			hasComponents = true
		}
	}
}

func newMainWorld() *world {
	w := &world{name: "main"}

	player := newPlayerEntity()
	draw := newDrawSystem().(*drawSystem)
	draw.cameraTarget = player
	w.addSystem(draw)
	w.addSystem(newPlayerSytem())
	w.addSystem(newUISystem())
	w.addEntity(player)
	w.addEntity(newBoxEntity())

	w.updateSystems()
	w.sortEntities()
	return w
}

File: library/world.lua
return function(name)
    local w = {}
    w.name = name
    w.systems = {}
    w.entities = {}

    w.addEntity = function(entity)
        w.entities[#w.entities + 1] = entity
    end

    w.addSystem = function(system)
        w.systems[#w.systems + 1] = system
    end

    w.sortEntities = function()
        for index = 1, #w.systems do
            w.systems[index].sortEntities()
        end
    end

    w.updateSystems = function()
        local system
        local entity
        local requiredComponent
        local component
        local hasComponent = false
        local hasAllComponents = false
        for systemIndex = 1, #w.systems do
            system = w.systems[systemIndex]
            for entityIndex = 1, #w.entities do
                entity = w.entities[entityIndex]
                for requiredComponentIndex = 1, #system.components do
                    requiredComponent = system.components[requiredComponentIndex]
                    if entity[requiredComponent] then
                        hasComponent = true
                    end
                    hasAllComponents = hasComponent
                    hasComponent = false
                end
                if hasAllComponents then
                    system.addEntity(entity)
                end
                hasAllComponents = false
            end
        end
    end

    w.load = function(arg)
        for index = 1, #w.systems do
            w.systems[index].load(arg)
        end
    end

    w.update = function(dt)
        for index = 1, #w.systems do
            w.systems[index].update(dt)
        end
    end

    w.draw = function()
        for index = 1, #w.systems do
            w.systems[index].draw()
        end
    end

    w.keypressed = function(key)
        for index = 1, #w.systems do
            w.systems[index].keypressed(key)
        end
    end

    w.keyreleased = function(key)
        for index = 1, #w.systems do
            w.systems[index].keyreleased(key)
        end
    end

    w.resize = function(width, height)
        for index = 1, #w.systems do
            w.systems[index].resize(width, height)
        end
    end

    return w
end
